### PR TITLE
Userdata to display static NAT as public ip instead of VR ip

### PR DESCRIPTION
**Pull request by ravening**
## Description

If static nat is enabled on VM then metadata service should return
the static nat instead of gateway IP.
If static not is not enabled then it should return the gateway IP
as the public IP

## Types of changes

- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] New feature (non-breaking change which adds functionality)
- [X] Bug fix (non-breaking change which fixes an issue)
- [ ] Enhancement (improves an existing feature and functionality)
- [ ] Cleanup (Code refactoring and cleanup, that may add test cases)

## How Has This Been Tested?

Step to reproduce:

1. Create a vm
2. Ssh to vm.
3. Run the below command inside the vm
   wget http://\<VR public ip\>/latest/meta-data/public-ipv4

   Note down the output of the above command
4. Now acquire a new public and enable static NAT on that IP to this vm
5. Now run the same command mentioned above in the VM
   This should display the static NAT ip instead of VR public IP

Output:

Before enabling static nat

wget http://10.10.10.40/latest/meta-data/public-ipv4
$ cat public-ipv4
10.10.10.29

After enabling static nat

wget http://10.10.10.40/latest/meta-data/public-ipv4
$ cat public-ipv4
10.11.10.30

### CODE REVIEW

**Comment by rhtyd** [comment]
@blueorangutan package

**Comment by rhtyd** [comment]
if @ravening does decide to change to 4.13, we can change the milestone to 4.13.1 @DaanHoogland  - it's not exactly a bug fix is it?

**Comment by DaanHoogland** [comment]
It is marked as a bug and could be considered so by a user that has a static address assigned. The want to be visible by that address in the outside world, @rhtyd .

**Comment by rhtyd** [comment]
@ravening cc @weizhouapache - can you explain how the user-data would be used by the user VM? Could that cause regressions for VMs which were previously expecting the VR/gateway public IP instead?

**Comment by andrijapanicsb** [comment]
When static nat is configured for a VM - then it actually shows it's IP address clearly, so this is in line with the fix I believe:

**Comment by andrijapanicsb** [comment]
@blueorangutan test

**Comment by andrijapanicsb** [comment]
@blueorangutan test

**Comment by rhtyd** [comment]
@blueorangutan package

**Comment by ravening** [comment]
> @ravening cc @weizhouapache - can you explain how the user-data would be used by the user VM? Could that cause regressions for VMs which were previously expecting the VR/gateway public IP instead?

If VM's dont have static NAT enabled, it will still display VR gateway as public ip. Only when static NAT is enabled, it will display static NAT as public ip. I guess that should be the expected behavior and shouldnt break regression

**Comment by ravening** [comment]
> @ravening this looks great, thanks. Can we rebase this on 4.13, please?

@DaanHoogland rebased to 4.13

**Comment by rhtyd** [comment]
@blueorangutan package

**Comment by andrijapanicsb** [comment]
will test once I'm available for the sake of "somebody else tested as well" thx.

**Comment by rhtyd** [comment]
@blueorangutan test

**Comment by rhtyd** [comment]
Can anyone confirm manual testing? Are we breaking any compatibility?

**Comment by DaanHoogland** [comment]
@wido @GabrielBrascher Are using this kind of configuration? @pdion891 @swill ?

@rhtyd I don't see any related error in that report from the blue ape. I am fine with merging.

/cc @ravening @weizhouapache

**Comment by andrijapanicsb** [comment]
testing it now

**Comment by andrijapanicsb** [comment]
Tested fine, except for a single issue I noticed (might not be related, but it also might)
Happens in both VPC and Isolated network
- Configure Static NAT, check meta-data - correct Static NAT IP reported
- Release the IP, without previously disabling the Static NAT - and still the same Public IP is reported in meta-data, which is wrong (should be the source NAT now)
- FYI: Disabling the Static NAT works fine, again Source NAT is reported in meta-data

The issue has something to do with IPtables and the way we (don't) clean those rules properly, see the image below.

@weizhouapache if you think this is not related to the PR, then we can move forward with this one (and this is kind of edge case, you are usually supposed to disable Static NAT, but again - a valid issue anyway)

image

**Comment by weizhouapache** [comment]
> Tested fine, except for a single issue I noticed (might not be related, but it also might)
> Happens in both VPC and Isolated network
> 
> * Configure Static NAT, check meta-data - correct Static NAT IP reported
> * Release the IP, without previously disabling the Static NAT - and still the same Public IP is reported in meta-data, which is wrong (should be the source NAT now)
> * FYI: Disabling the Static NAT works fine, again Source NAT is reported in meta-data
> 
> The issue has something to do with IPtables and the way we (don't) clean those rules properly, see the image below.
> 
> @weizhouapache if you think this is not related to the PR, then we can move forward with this one (and this is kind of edge case, you are usually supposed to disable Static NAT, but again - a valid issue anyway)
> 
> image

@andrijapanicsb thanks for testing !
It is definitely an issue. We will fix it.

File: server/src/main/java/com/cloud/network/router/CommandSetupHelper.java [hunk 1]
```diff
@@ -1024,18 +1026,21 @@ private VmDataCommand generateVmDataCommand(final VirtualRouter router, final St
...
+        if (dcVo.getNetworkType() == NetworkType.Basic || network.getGuestType() == Network.GuestType.Shared) {
+            cmd.addVmData("metadata", "public-ipv4", vmPrivateIpAddress);
             cmd.addVmData("metadata", "public-hostname", StringUtils.unicodeEscape(vmName));
         } else {
-            if (router.getPublicIpAddress() == null) {
-                cmd.addVmData("metadata", "public-ipv4", guestIpAddress);
+            if (publicIpAddress != null) {
+                cmd.addVmData("metadata", "public-ipv4", publicIpAddress);
+                cmd.addVmData("metadata", "public-hostname", publicIpAddress);
             } else {
```
Comment (weizhouapache) [inline server/src/main/java/com/cloud/network/router/CommandSetupHelper.java]: @ravening you missed one check here
```
-            } else {
+            } else if (router.getPublicIpAddress() != null) {
```

Comment (ravening) [reply server/src/main/java/com/cloud/network/router/CommandSetupHelper.java]: Made necessary chagnes

**Comment by ravening** [comment]
> Tested fine, except for a single issue I noticed (might not be related, but it also might)
> Happens in both VPC and Isolated network
> 
> * Configure Static NAT, check meta-data - correct Static NAT IP reported
> * Release the IP, without previously disabling the Static NAT - and still the same Public IP is reported in meta-data, which is wrong (should be the source NAT now)
> * FYI: Disabling the Static NAT works fine, again Source NAT is reported in meta-data
> 
> The issue has something to do with IPtables and the way we (don't) clean those rules properly, see the image below.
> 
> @weizhouapache if you think this is not related to the PR, then we can move forward with this one (and this is kind of edge case, you are usually supposed to disable Static NAT, but again - a valid issue anyway)
> 
> image

@andrijapanicsb Thanks for testing. I have updated the commit. Can you please test it again and let me know if you face any issues?

**Comment by weizhouapache** [comment]
@ravening the new commits is not related to the issue @andrijapanicsb reported.
can you test following what @andrijapanicsb said ?

```
Release the IP, without previously disabling the Static NAT - and still the same Public IP is reported in meta-data, which is wrong (should be the source NAT now)
```

**Comment by ravening** [comment]
> Release the IP, without previously disabling the Static NAT - and still the same Public IP is reported in meta-data, which is wrong (should be the source NAT now)

@weizhouapache tested the steps which @andrijapanicsb mentioned above.
Yes it does report the public static nat ip instead of source nat ip if I release an IP without disabling the static nat.

**Comment by rhtyd** [comment]
@blueorangutan package

**Comment by rhtyd** [comment]
@blueorangutan test

**Comment by DaanHoogland** [comment]
@andrijapanicsb can you re-test?

**Comment by rhtyd** [comment]
@andrijapanicsb can you re-test?

**Comment by andrijapanicsb** [comment]
incoming...

**Comment by andrijapanicsb** [comment]
Merging based on 3 x LGTMs/Approvals and extensive manual testing, and passing regression tests.

**Pull request merged**